core/src/service/sync_service.rs

```diff
@@ -24,122 +28,251 @@ pub struct WorkCalculated {
+    public_key_by_username: &'a mut TransactionTable<'b, String, Owner, PublicKeyByUsername>,
+
+    // account pre-loaded for convenience
+    account: Account,
+
+    // client for network requests and config for disk i/o
+    client: &'a Client,
+    config: &'a Config,
+}
+
+impl<'a, 'b, Base, Local, UsernameByPublicKey, PublicKeyByUsername, Client>
+    SyncContext<'a, 'b, Base, Local, UsernameByPublicKey, PublicKeyByUsername, Client>
+where
+    Base: TreeLike<F = SignedFile>,
+    Local: TreeLike<F = SignedFile>,
+    UsernameByPublicKey: SchemaEvent<Owner, String>,
+    PublicKeyByUsername: SchemaEvent<String, Owner>,
+    Client: Requester,
+{
+    fn sync<F>(&mut self, report_sync_operation: &mut F) -> CoreResult<i64>
+    where
+        F: FnMut(SyncOperation),
+    {
+        self.prune()?;
+        let update_as_of = self.pull(report_sync_operation)?;
+        self.last_synced = Some(update_as_of as u64);
+        self.push_metadata(report_sync_operation)?;
+        self.push_documents(report_sync_operation)?;
+        Ok(update_as_of)
     }
 
     /// Pulls remote changes and constructs a changeset Merge such that Stage<Stage<Stage<Base, Remote>, Local>, Merge> is valid.
     /// Promotes Base to Stage<Base, Remote> and Local to Stage<Local, Merge>
-    fn pull<F>(&mut self, update_sync_progress: &mut F) -> CoreResult<i64>
+    fn pull<F>(&mut self, report_sync_operation: &mut F) -> CoreResult<i64>
     where
-        F: FnMut(SyncProgressOperation),
+        F: FnMut(SyncOperation),
     {
         // fetch metadata updates
-        update_sync_progress(SyncProgressOperation::StartWorkUnit(ClientWorkUnit::PullMetadata));
-        let updates = self.get_updates()?;
-        let mut remote_changes = updates.file_metadata;
-        let update_as_of = updates.as_of_metadata_version;
+        let (mut remote_changes, update_as_of) = {
+            report_sync_operation(SyncOperation::PullMetadataStart);
+
+            let updates = self.get_updates()?;
+            let mut remote_changes = updates.file_metadata;
+            let update_as_of = updates.as_of_metadata_version;
+
+            remote_changes = self.prune_remote_orphans(remote_changes)?;
+            self.populate_public_key_cache(&remote_changes)?;
+
+            let mut remote = self.base.stage(remote_changes).to_lazy();
+            report_sync_operation(SyncOperation::PullMetadataEnd(
+                remote.resolve_and_finalize_all(
+                    &self.account,
+                    remote.tree.staged.owned_ids().into_iter(),
+                    self.username_by_public_key,
+                )?,
+            ));
+            let (_, remote_changes) = remote.unstage();
+            (remote_changes, update_as_of)
+        };
 
         // initialize root if this is the first pull on this device
-        if self.tx.root.get(&OneKey {}).is_none() {
+        if self.root.is_none() {
```

Comment:
use local fields of new struct instead of actual tables so that changes can be read by later parts of sync and optionally applied